crates/block-producer/src/state_view/account_state.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::{BTreeMap, VecDeque};
+
+use miden_objects::{accounts::AccountId, Digest};
+
+/// Tracks the account state transitions resulting from inflight transactions.
+///
+/// New transactions can be registered with [Self::add]. States that are no longer considered inflight
+/// (e.g. due to being applied) may be removed using [Self::remove].
+///
+/// Both functions perform safety checks to ensure the states match what we expect.
+#[derive(Debug, Default)]
+pub struct InflightAccountStates(BTreeMap<AccountId, VecDeque<Digest>>);
+
+impl InflightAccountStates {
+    /// Push a new inflight account state transition.
+    ///
+    /// `init_state` should match the previous `final_state`. Error contains
+    /// this value if the condition is not met.
+    pub fn add(
+        &mut self,
+        id: AccountId,
+        init_state: Digest,
+        final_state: Digest,
+    ) -> Result<(), Digest> {
+        debug_assert_ne!(final_state, Digest::default(), "Final state should not be null");
+
+        // New accounts have a default state.
+        let current_state =
+            self.0.get(&id).and_then(|states| states.back()).copied().unwrap_or_default();
+
+        if current_state != init_state {
+            return Err(current_state);
+        }
+
+        self.0.entry(id).or_default().push_back(final_state);
```

Comment:
Is this logic correct? What if we try to add an update for an existing account? That is, the `init_state` is not empty but the internal map has no entry for this account.

I think the high-level logic should be:
- If the entry for the account is already in the map, make sure the last state is the same as `init_state` and then append the `final_state` to the queue.
- If the entry is not in the map yet, insert a new entry for the account id with `final_state` in the single element in the queue.

Reply:
Ah. Right there is no way to test for an empty initial state because this is local only and excludes state that could be in the store.

Thanks!